Framework/NexusGeometry/src/NexusGeometrySave.cpp

```diff
@@ -0,0 +1,805 @@
+    writeStrAttribute(yPixelOffset, UNITS, METRES);
+  }
+
+  if (!zIsZero) {
+    zPixelOffset =
+        grp.createDataSet(Z_PIXEL_OFFSET, H5::PredType::NATIVE_DOUBLE, space);
+    zPixelOffset.write(posz.data(), H5::PredType::NATIVE_DOUBLE);
+    writeStrAttribute(zPixelOffset, UNITS, METRES);
+  }
+}
+
+/*
+ * Function: writeNXDetectorNumber
+ * For use with NXdetector group. Writes the detector numbers for all detector
+ * pixels in compInfo to a new dataset in the group.
+ *
+ * @param detectorIDs : std::vector<int> container of all detectorIDs to be
+ * stored into dataset 'detector_number'.
+ * @idx : size_t index of bank in compInfo.
+ */
+void writeNXDetectorNumber(H5::Group &grp,
+                           const Geometry::ComponentInfo &compInfo,
+                           const std::vector<int> &detectorIDs,
+                           const size_t idx) {
+
+  H5::DataSet detectorNumber;
+
+  std::vector<int> bankDetIDs; // IDs of detectors beloning to bank
+  std::vector<size_t> bankDetectors = compInfo.detectorsInSubtree(idx);
+  bankDetIDs.reserve(bankDetectors.size());
+
+  for (size_t &index : bankDetectors) {
+    bankDetIDs.push_back(detectorIDs[index]);
+  }
+
+  const auto nDetectorsInBank = static_cast<hsize_t>(bankDetIDs.size());
+
+  int rank = 1;
+  hsize_t dims[static_cast<hsize_t>(1)];
+  dims[0] = nDetectorsInBank;
+
+  H5::DataSpace space = H5Screate_simple(rank, dims, NULL);
+
+  detectorNumber =
+      grp.createDataSet(DETECTOR_IDS, H5::PredType::NATIVE_INT, space);
+  detectorNumber.write(bankDetIDs.data(), H5::PredType::NATIVE_INT, space);
+}
+
+/*
+ * Function: writeNXDetectorNumber
+ * For use with NXmonitor group. write 'detector_id's of an NXmonitor, which is
+ * a specific type of NXdetector, to its group.
+ *
+ * @param grp : NXmonitor group (HDF group)
+ * @param compInfo : componentInfo object.
+ * @param monitorIDs : std::vector<int> container of all monitorIDs to be stored
+ * into dataset 'detector_id' (or 'detector_number'. naming convention
+ * inconsistency?).
+ * @idx : size_t index of monitor in compInfo.
+ */
+void writeNXMonitorNumber(H5::Group &grp,
+                          const Geometry::ComponentInfo &compInfo,
+                          const std::vector<int> &monitorIDs,
+                          const size_t idx) {
+
+  // these DataSets are duplicates of each other. written to the NXmonitor group
+  // to handle the naming inconsistency. probably temporary.
+  H5::DataSet detectorNumber, detector_id;
+
+  std::vector<int> monitorDetIDs;
+  std::vector<size_t> bankDetectors = compInfo.detectorsInSubtree(idx);
+  monitorDetIDs.reserve(bankDetectors.size());
+
+  for (size_t &index : bankDetectors) {
+    monitorDetIDs.push_back(monitorIDs[index]);
+  }
+
+  const auto nMonitorsInBank = static_cast<hsize_t>(monitorDetIDs.size());
+
+  int rank = 1;
+  hsize_t dims[static_cast<hsize_t>(1)];
+  dims[0] = nMonitorsInBank;
+
+  H5::DataSpace space = H5Screate_simple(rank, dims, NULL);
```

Comment:
`nullptr`, and several more cases in this file.

Reply:
done